### PR TITLE
Improvement/136 extraction pipeline

### PR DESCRIPTION
**Pull request by nils-schmitt**

### CODE REVIEW

File: tracex_project/extraction/logic/module.py [hunk 1]
```diff
@@ -27,31 +28,38 @@ def execute(
...
     ) -> pd.DataFrame:
         """
         Executes the logic of the module. Override this to define your own module.
-        Every module receives the patient journey as parameter which is set to the instance variable of each module.
-        This method should always return a dataframe, so other modules can use the result.
+
+        Keyword arguments:
+        _input -- Any additional input to the module
+        patient_journey -- The patient journey as text
+        patient_journey_sentences -- The same patient journey as a list of sentences
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/module.py:41]:
```suggestion
        _input -- Any additional input to the module.
        patient_journey -- The patient journey as text.
        patient_journey_sentences -- The same patient journey as a list of sentences.
```

File: tracex_project/extraction/logic/modules/module_activity_labeler.py [hunk 2]
```diff
@@ -23,8 +24,8 @@ def __init__(self):
     def execute(
         self,
         _input=None,
-        patient_journey=None,
-        patient_journey_sentences=None,
+        patient_journey: str = None,
+        patient_journey_sentences: List[str] = None,
         cohort=None,
     ):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_activity_labeler.py:30]:
```suggestion
    ) -> pd.DataFrame:
```

File: tracex_project/extraction/logic/modules/module_activity_labeler.py [hunk 3]
```diff
@@ -64,24 +65,18 @@ def __number_patient_journey_sentences(patient_journey_sentences):
         return patient_journey_numbered
 
     @staticmethod
-    def __extract_activities(patient_journey_numbered, condition):
+    def __extract_activities(patient_journey_numbered: str, condition: Optional[str]):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_activity_labeler.py:68]:
```suggestion
    def __extract_activities(patient_journey_numbered: str, condition: Optional[str]) -> pd.DataFrame:
```

File: tracex_project/extraction/logic/modules/module_activity_labeler.py [hunk 4]
```diff
@@ -39,19 +40,19 @@ def execute(
...
         )
 
-        condition = cohort["condition"] if cohort is not None else None
+        condition = getattr(cohort, "condition", None)
 
-        patient_journey_numbered = self.__number_patient_journey_sentences(
+        patient_journey_numbered: str = self.__number_patient_journey_sentences(
             patient_journey_sentences
         )
         activity_labels = self.__extract_activities(patient_journey_numbered, condition)
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_activity_labeler.py:48]:
```suggestion
        activity_labels: pd.DataFrame = self.__extract_activities(patient_journey_numbered, condition)
```

File: tracex_project/extraction/logic/modules/module_event_type_classifier.py [hunk 5]
```diff
@@ -23,18 +24,21 @@ def __init__(self):
 
     @log_execution_time(Path(settings.BASE_DIR / "tracex/logs/execution_time.log"))
     def execute(
-        self, df, patient_journey=None, patient_journey_sentences=None, cohort=None
+        self,
+        df: pd.DataFrame,
+        patient_journey=None,
+        patient_journey_sentences=None,
+        cohort=None,
     ):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_event_type_classifier.py:32]:
```suggestion
    ) -> pd.DataFrame:
```

File: tracex_project/extraction/logic/modules/module_location_extractor.py [hunk 6]
```diff
@@ -21,18 +22,21 @@ def __init__(self):
 
     @log_execution_time(Path(settings.BASE_DIR / "tracex/logs/execution_time.log"))
     def execute(
-        self, df, patient_journey=None, patient_journey_sentences=None, cohort=None
+        self,
+        df: pd.DataFrame,
+        patient_journey=None,
+        patient_journey_sentences=None,
+        cohort=None,
     ):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_location_extractor.py:30]:
```suggestion
    ) -> pd.DataFrame:
```

File: tracex_project/extraction/logic/modules/module_metrics_analyzer.py [hunk 7]
```diff
@@ -26,10 +26,14 @@ def __init__(self):
 
     @log_execution_time(Path(settings.BASE_DIR / "tracex/logs/execution_time.log"))
     def execute(
-        self, df, patient_journey=None, patient_journey_sentences=None, cohort=None
+        self,
+        df: pd.DataFrame,
+        patient_journey=None,
+        patient_journey_sentences=None,
+        cohort=None,
     ):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_metrics_analyzer.py:34]:
```suggestion
    ) -> pd.DataFrame:
```

File: tracex_project/extraction/logic/modules/module_metrics_analyzer.py
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_metrics_analyzer.py:1]: Please add type hints to rate_activity_revelance and rate_timestamp_correctness

File: tracex_project/extraction/logic/modules/module_patient_journey_preprocessor.py
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_patient_journey_preprocessor.py:1]: Please add type hints

File: tracex_project/extraction/logic/modules/module_time_extractor.py [hunk 8]
```diff
@@ -22,9 +23,14 @@ def __init__(self):
 
     @log_execution_time(Path(settings.BASE_DIR / "tracex/logs/execution_time.log"))
     def execute(
-        self, df, patient_journey=None, patient_journey_sentences=None, cohort=None
+        self,
+        df: pd.DataFrame,
+        patient_journey=None,
+        patient_journey_sentences: List[str] = None,
+        cohort=None,
     ):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_time_extractor.py:31]:
```suggestion
    ) -> pd.DataFrame:
```

File: tracex_project/extraction/logic/modules/module_time_extractor.py
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_time_extractor.py:1]: Please add type hints to all functions in this file

File: tracex_project/extraction/logic/orchestrator.py [hunk 9]
```diff
@@ -134,39 +160,39 @@ def initialize_modules(self):
             key: self.get_configuration().modules[key]()
             for key in self.get_configuration().modules
         }
-        print("Initialization of modules successful.")
         return modules
 
     def run(self, view=None):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/orchestrator.py:165]:
```suggestion
    def run(self, view=None) -> pd.DataFrame:
```

File: tracex_project/tracex/logic/utils.py [hunk 10]
```diff
@@ -277,3 +290,15 @@ def set_default_timestamps(df):
             df["time:end_timestamp"], format="%Y%m%dT%H%M", errors="coerce"
         )
         df["time:duration"] = "00:01:00"
+
+    @staticmethod
+    def delete_metrics_columns(df: pd.DataFrame):
```
Comment (PitButtchereit) [inline tracex_project/tracex/logic/utils.py:295]:
```suggestion
    def delete_metrics_columns(df: pd.DataFrame) -> pd.DataFrame:
```

File: tracex_project/tracex/logic/utils.py [hunk 11]
```diff
@@ -206,6 +207,18 @@ def dataframe_to_xes(df, name, activity_key):
 
         return file_path
 
+    @staticmethod
+    def text_to_sentence_list(text):
```
Comment (PitButtchereit) [inline tracex_project/tracex/logic/utils.py:211]:
```suggestion
    def text_to_sentence_list(text: str) -> List[str]:
```

**Review by PitButtchereit — CHANGES_REQUESTED** [review]

File: tracex_project/extraction/logic/modules/module_metrics_analyzer.py [hunk 12]
```diff
@@ -56,7 +61,7 @@ def execute(
         return metrics_df
 
     @staticmethod
-    def __rate_activity_relevance(activity, condition):
+    def __rate_activity_relevance(activity, condition) -> str:
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_metrics_analyzer.py:64]: Please add type hints to the function arguments

File: tracex_project/extraction/logic/modules/module_metrics_analyzer.py [hunk 13]
```diff
@@ -86,7 +91,7 @@ def __rate_activity_relevance(activity, condition):
 
         return category
 
-    def __rate_timestamps_correctness(self, activity, start, end):
+    def __rate_timestamps_correctness(self, activity, start, end) -> Tuple[str, float]:
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_metrics_analyzer.py:94]: Please add type hints to the function arguments

File: tracex_project/extraction/logic/modules/module_time_extractor.py [hunk 14]
```diff
@@ -91,9 +97,27 @@ def __calculate_duration(row):
...
+                df[column] = df[column].fillna(pd.Timestamp("2020-01-01 00:00"))
+
+            return df
+
+        def fill_missing_values(df, column):
+            df[column] = df[column].ffill().bfill()
+
+            return df
+
         def fix_end_dates(row):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/modules/module_time_extractor.py:121]: Please add type hints here

File: tracex_project/extraction/logic/modules/module_time_extractor.py [hunk 15]
```diff
@@ -91,9 +97,27 @@ def __calculate_duration(row):
         return f"{int(hours):02d}:{int(minutes):02d}:{int(seconds):02d}"
 
     @staticmethod
-    def __post_processing(df):
+    def __post_processing(df: pd.DataFrame) -> pd.DataFrame:
         """Fill missing values for dates with default values."""
 
+        def convert_to_datetime(df, column):
```
Comment (FR-SON) [inline tracex_project/extraction/logic/modules/module_time_extractor.py:103]:
```suggestion
        def convert_to_datetime(df: pd.DataFrame, column: str) -> pd.DataFrame:
```

File: tracex_project/extraction/logic/modules/module_time_extractor.py [hunk 16]
```diff
@@ -91,9 +97,27 @@ def __calculate_duration(row):
...
+            df[column] = pd.to_datetime(
+                df[column], format="%Y%m%dT%H%M", errors="coerce"
+            )
+
+            return df
+
+        def set_default_date_if_na(df, column):
```
Comment (FR-SON) [inline tracex_project/extraction/logic/modules/module_time_extractor.py:110]:
```suggestion
        def set_default_date_if_na(df: pd.DataFrame, column: str) -> pd.DataFrame:
```

File: tracex_project/extraction/logic/modules/module_time_extractor.py [hunk 17]
```diff
@@ -91,9 +97,27 @@ def __calculate_duration(row):
...
+            if df[column].isna().all():
+                df[column] = df[column].fillna(pd.Timestamp("2020-01-01 00:00"))
+
+            return df
+
+        def fill_missing_values(df, column):
```
Comment (FR-SON) [inline tracex_project/extraction/logic/modules/module_time_extractor.py:116]:
```suggestion
        def fill_missing_values(df: pd.DataFrame, column: str) -> pd.DataFrame:
```

File: tracex_project/extraction/logic/orchestrator.py [hunk 18]
```diff
@@ -112,17 +140,15 @@ def set_db_objects_id(self, object_name: str, object_id: int):
         """Set the database id objects for the orchestrator instance."""
         self.db_objects_id[object_name] = object_id
 
-    def get_db_objects_id(self, object_name):
+    def get_db_objects_id(self, object_name: str):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/orchestrator.py:143]: Can you add a type hint for what this method returns

File: tracex_project/extraction/logic/orchestrator.py [hunk 19]
```diff
@@ -112,17 +140,15 @@ def set_db_objects_id(self, object_name: str, object_id: int):
...
         return self.db_objects_id[object_name]
 
-    def update_modules(self, modules_list):
-        """Update the modules of the orchestrator instance."""
-        modules_dictionary = self.get_configuration().modules
-        updated_modules = {
-            key: modules_dictionary[key]
-            for key in modules_dictionary
-            if key in modules_list
+    def reduce_modules_to(self, modules: List):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/orchestrator.py:147]: please add type hint for return value

File: tracex_project/extraction/logic/modules/module_time_extractor.py (hunk 14 above)
Comment (FR-SON) [inline tracex_project/extraction/logic/modules/module_time_extractor.py:121]:
```suggestion
        def fix_end_dates(row: pd.Series) -> pd.Series:
```

File: tracex_project/extraction/logic/orchestrator.py [hunk 20]
```diff
@@ -226,7 +252,7 @@ def save_results_to_db(self):
         patient_journey.save()
 
     def set_default_values(self):
```
Comment (PitButtchereit) [inline tracex_project/extraction/logic/orchestrator.py:254]:
```suggestion
    def set_default_values(self) -> None:
```

**Review by PitButtchereit — CHANGES_REQUESTED** [review]

File: tracex_project/extraction/logic/orchestrator.py [hunk 21]
```diff
@@ -213,7 +239,7 @@ def save_results_to_db(self):
             events_with_metric_list.append(event)
             metric_list.append(metric)
 
-        events: List[Event] = Event.manager.bulk_create(events_with_metric_list)
+        events: list[Event] = Event.manager.bulk_create(events_with_metric_list)
```
Comment (FR-SON) [inline tracex_project/extraction/logic/orchestrator.py:242]:
```suggestion
        events: List[Event] = Event.manager.bulk_create(events_with_metric_list)
```

File: tracex_project/extraction/views.py [hunk 22]
```diff
@@ -183,7 +219,22 @@ def get_form_kwargs(self):
...
+        Args:
+            **kwargs: Arbitrary keyword arguments.
+
+        Returns:
+            dict: The context data for the result page.
+        """
+
         context = super().get_context_data(**kwargs)
         orchestrator = Orchestrator.get_instance()
         activity_key = orchestrator.get_configuration().activity_key
```
Comment (FR-SON) [inline tracex_project/extraction/views.py:240]:
```suggestion
        activity_key: str = orchestrator.get_configuration().activity_key
```

File: tracex_project/extraction/views.py [hunk 23]
```diff
@@ -223,17 +274,22 @@ def get_context_data(self, **kwargs):
         return context
 
     @staticmethod
-    def build_trace_df(filter_dict):
-        """Build the trace df based on the filter settings."""
+    def build_trace_df(filter_dict: Dict[str, List[str]]) -> pd.DataFrame:
+        """Build the trace dataframe based on the filter settings."""
         trace_df = Orchestrator.get_instance().get_data()
```
Comment (FR-SON) [inline tracex_project/extraction/views.py:279]:
```suggestion
        trace_df: pd.DataFrame = Orchestrator.get_instance().get_data()
```

**Review by FR-SON — CHANGES_REQUESTED** [review]

**Review by PitButtchereit — CHANGES_REQUESTED** [review]

**Review by FR-SON — APPROVED** [review]

**Comment by nils-schmitt** [comment]
You can push, as soon as you have approved

**Review by PitButtchereit — APPROVED** [review]
I am happy 🙂

**Pull request merged**